_episodes/l1-05-test_doubles.md

```diff
@@ -0,0 +1,238 @@
+
+```cpp
+class Animal {
+  virtual ~Animal() {};
+  virtual double walk(int steps);
+  virtual void eat(double carbs);
+  virtual void die();
+};
+```
+
+The corresponding mocked class will be:
+
+```cpp
+class MockAnimal : public Animal {
+ public:
+  MOCK_METHOD(double, walk, (int), (override));
+  MOCK_METHOD(void, eat, (double), (override));
+  MOCK_METHOD(void, die, (), (override));
+};
+```
+
+Now let's write a test for the following function, which finds out if an animal is dead or alive at the end of the day depending on how much food it has taken and how much it has walked.
+
+```cpp
+bool is_alive_at_end_of_day(int steps, double carbs, Animal animal) {
+  double spent_carbs{animal.walk(steps)};
+  if (spent_carbs > carbs) {
+    animal.die();
+    return false;
+  }
+  animal.eat(carbs - spent_carbs);
+  return true;
+}
+
+If we were to use a real implementation of `Animal`, let's say a `Horse`, testing this function would be complicated because the result would depend on the specific metabolism of the animal, which might be quite complicated (and potentially time consuming to run). So we can use `MockAnimal` instead to ensure that the logic of the function is correct. A couple of tests for this would look as:
+
+```cpp
+using ::testing::Return;
+
+TEST(IsAliveTest, Lives) {
+  Animal animal = MockAnimal();
+  int steps{400};
+  double carbs{2000.0};
+  double consumed{500.0};
+
+  EXPECT_CALL(animal, walk(steps)).Times(1).WillOnce(Return(consumed));
+  EXPECT_CALL(animal, eat(carbs-consumed)).Times(1);
+  EXPECT_CALL(animal, die()).Times(0);
+  ASSERT_TRUE(is_alive_at_end_of_day(steps, carbs, animal))
+}
+
+TEST(IsAliveTest, Dies) {
+  Animal animal = MockAnimal();
+  int steps{400};
+  double carbs{2000.0};
+  double consumed{5000.0};
+
+  EXPECT_CALL(animal, walk(steps)).Times(1).WillOnce(Return(consumed));
+  EXPECT_CALL(animal, eat(carbs-consumed)).Times(1);
+  EXPECT_CALL(animal, die()).Times(1);
+  ASSERT_FALSE(is_alive_at_end_of_day(steps, carbs, animal))
+}
+```
+
+[Google Mock]: https://google.github.io/googletest/gmock_for_dummies.html
+[Mocking Cookbook]: https://google.github.io/googletest/gmock_cook_book.html
+
+> ### Mocking is not always the solution
+>
+> In the above example, it would have been tricky to test the logic of the function in full without mocks. However, they are not always the solution. Mocks do not work with  top level functions, only with classes. Depending on the complexity of the class, setting up the mock might be too complicated and not worth it for testing the function of interest. Very often, stubs, fakes and dummies will carry you a long way before you need to use mocks.
```

Comment:
```suggestion
> In the above example, it would have been tricky to test the logic of the function in full without mocks. However, they are not always the solution. Mocks do not work with top level functions, only with classes. Depending on the complexity of the class, setting up the mock might be too complicated and not worth it for testing the function of interest. Very often, stubs, fakes and dummies will carry you a long way before you need to use mocks.
```